### PR TITLE
521-refactor: Fsd widget communication

### LINKED ISSUE #521: FSD: Widget Communication
## 📝 Describe the refactoring task

Part of the issue #347 

## 📍 Code location

`src/widgets/communication`

## ✅ Definition of Done (DoD):

1. Refactor `communication.tsx`'s `scss` to `scss modules`

-   remove all parent selector `& .card-content` -> `.card-content`. 
-   change all possible values to variables if such exists.
-   and everything in your opinion that needs to be refactored

#### `communication.tsx`
1. rename `RequiredProps` to `CommunicationProps`
2. change all the `interfaces` to `types`
3. remove discord logo wrapper (use `Image` component directly)
4. move `communication.test.tsx` to `ui` folder
5. change test to not render `Communication` component with the same course on every test case.

### PR DESCRIPTION
**Pull request by thirdmadman**
## What type of PR is this? (select all that apply)

- [ ] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 🚧 Breaking Change
- [x] 🧑‍💻 Code Refactor
- [ ] 📝 Documentation Update

## Description

- [x] 1. Refactor `communication.tsx`'s `scss` to `scss modules`
- [x] -   remove all parent selector `& .card-content` -> `.card-content`. 
- [x] -   change all possible values to variables if such exists.
- [x] -   and everything in your opinion that needs to be refactored
#### `communication.tsx`
- [x] 1. rename `RequiredProps` to `CommunicationProps`
- [x] 2. change all the `interfaces` to `types`
- [x] 3. remove discord logo wrapper (use `Image` component directly)
- [x] 4. move `communication.test.tsx` to `ui` folder
- [x] 5. change test to not render `Communication` component with the same course on every test case.

## Related Tickets & Documents

- Related Issue #521 
- Closes #521 #347

## Screenshots, Recordings

none

## Added/updated tests?

- [x] 👌 Yes
- [ ] 🙅‍♂️ No, because they aren't needed
- [ ] 🙋‍♂️ No, because I need help

## Summary by CodeRabbit

- **New Features**
	- Introduced a new `communicationText` export for localized communication content in English and Russian.
	- Added a new SCSS file for styling the communication module.

- **Bug Fixes**
	- Improved rendering logic in the `Communication` component for better maintainability.

- **Refactor**
	- Simplified the props structure in the `Communication` component.
	- Updated the rendering logic to utilize centralized text data.
	- Replaced the previous Discord logo component with a new Image component for improved styling.

- **Tests**
	- Added unit tests for the `Communication` component to ensure proper rendering of localized content.

### CODE REVIEW

File: src/widgets/communication/ui/locales.ts
Comment (Quiddlee) [inline src/widgets/communication/ui/locales.ts:1]: we don't use `locales.ts` file in our codebase. Better use consistent `constants.ts` file

Comment (thirdmadman) [reply src/widgets/communication/ui/locales.ts:1]: Sure, we don't. 
For now, as you can see from the screenshot, we store localies right in the component files.

Also, we use separate data files like ```react-ru.data.ts``` and  ```react-en.data.ts```, but this is kind of a different pattern.

But Im not sure where we use constants.ts for localies, explicitly with strucure like 

```tsx
const localizedContent = {
  en: {
  ...
  },
  ru: {
  ...
  }
};
```

Maybe we can discuss this moment in chat?

image

Comment (thirdmadman) [reply src/widgets/communication/ui/locales.ts:1]: fixed in: https://github.com/rolling-scopes/site/pull/536/commits/12a3a65d26d47cff21bad6056bf09253fad23b33

**Review by Quiddlee — APPROVED** [review]

File: src/widgets/communication/ui/communication.module.scss [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+.discord-logo-wrapper {
+  flex-shrink: 0;
+  align-self: center;
+
+  width: 250px;
+  padding: 30px;
+
+  background-color: #5865f2;
```
[the remark this answers is not in the document]
Comment (thirdmadman) [reply src/widgets/communication/ui/communication.module.scss:32]: Colors in the format hsl is used only in ```_constants.scss``` file, but hex form of colors is used in other ```*.module.scss``` and ```*.scss``` files.

If it's an anomaly and there should be used other color form constants - tell me which one, I will replace.

File: dev-data/widget-communication.data.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
+export const communicationText = {
```
Comment (SpaNb4) [inline dev-data/widget-communication.data.ts:1]: We need to come up with some naming rules for cases like this https://github.com/rolling-scopes/site/pull/538#discussion_r1773365305

File: src/widgets/communication/ui/communication.test.tsx [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    const subtitleElement = screen.getByText(subTitle);
+    const firstParagraphElement = screen.getByText(`${firstParagraphFirstHalf}`, { exact: false });
+    const linkElement = screen.getByText(discordLink);
+
+    expect(titleElement).toBeVisible();
+    expect(subtitleElement).toBeVisible();
+    expect(firstParagraphElement).toBeVisible();
+    expect(linkElement).toBeVisible();
+    expect(linkElement.getAttribute('href')).toMatch(DISCORD_LINKS[firstCourse]);
+  });
```
Comment (SpaNb4) [inline src/widgets/communication/ui/communication.test.tsx:51]: Can it be "combined" like this https://github.com/rolling-scopes/site/pull/540#discussion_r1773397202?

File: src/widgets/communication/ui/communication.test.tsx [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    const firstCourse = Object.keys(DISCORD_LINKS)[0] as keyof typeof DISCORD_LINKS;
+
+    const {
+      title,
+      subTitle,
+      firstParagraphFirstHalf,
+      discordLink,
+    } = communicationText.en;
+
+    renderWithRouter(<Communication courseName={firstCourse} />);
```
Comment (SpaNb4) [inline src/widgets/communication/ui/communication.test.tsx:17]: I would just use the string `'react'` or whatever for `courseName`

Comment (thirdmadman) [reply src/widgets/communication/ui/communication.test.tsx:17]: I prefer to not rely on plain string. In this case, ```courseName``` is value from list of const. List and its value can change, but if the list persists, it's safe to use the first permitted value from this list.

File: dev-data/widget-communication.data.ts (hunk 2 above)
Comment (thirdmadman) [reply dev-data/widget-communication.data.ts:1]: Sure. Don't know how to name this better, but as long, as our next step is to refactor dev-data, I suppose any "ok" name fits.

File: src/widgets/communication/ui/communication.test.tsx (hunk 3 above)
Comment (thirdmadman) [reply src/widgets/communication/ui/communication.test.tsx:51]: Yeah, I understand what about you, but only for 2 test cases, this feels little to much, and is easier to just copy.
What do you think?

File: src/widgets/communication/ui/communication.module.scss (hunk 1 above)
[the remark this answers is not in the document]
Comment (ansivgit) [reply src/widgets/communication/ui/communication.module.scss:32]: we're trying to change all of them to constant or `hsl` format. You could try to choose more suitable constant here [pallete](https://www.figma.com/design/rF0bjhIoxrIw2GFOeRvxEU/RS?node-id=1-460&t=JHY6NtbphZlAOM2N-1)

File: src/widgets/communication/ui/communication.tsx [hunk 5]
```diff
@@ -9,81 +10,63 @@ import {
...
+
   return (
-    <section className="communication container">
-      <article className="communication content info-wrapper">
-        <WidgetTitle mods="asterisk">{localizedContent[lang].title}</WidgetTitle>
-        <div className="column-2">
-          <figure className="disclogo-wrapper">
-            <DiscordLogo />
+    <section className={cx('container')}>
+      <article className={cx('content', 'communication-content')}>
```
[the remark this answers is not in the document]
Comment (ansivgit) [reply src/widgets/communication/ui/communication.tsx:42]: I mean `article` ))

File: src/widgets/communication/ui/communication.tsx [hunk 6]
```diff
@@ -9,81 +10,63 @@ import {
...
+        <WidgetTitle mods="asterisk">{title}</WidgetTitle>
+        <div className={cx('communication-text')}>
```
Comment (ansivgit) [inline src/widgets/communication/ui/communication.tsx:44]:
```suggestion
        <div className={cx('communication-content')}>
```

File: src/widgets/communication/ui/communication.tsx [hunk 7]
```diff
@@ -9,81 +10,63 @@ import {
...
+          <figure className={cx('discord-logo-wrapper')}>
+            <Image src={discordLogo} alt="discord logo" />
           </figure>
           <div>
```
Comment (ansivgit) [inline src/widgets/communication/ui/communication.tsx:48]:
```suggestion
          <article className={cx('communication-description')}>
```

File: src/widgets/communication/ui/communication.test.tsx (hunk 3 above)
Comment (SpaNb4) [reply src/widgets/communication/ui/communication.test.tsx:51]: We can leave it as is for now and discuss it during our next call

File: src/widgets/communication/ui/communication.tsx (hunk 6 above)
Comment (thirdmadman) [reply src/widgets/communication/ui/communication.tsx:44]: image

File: src/widgets/communication/ui/communication.tsx (hunk 7 above)
Comment (thirdmadman) [reply src/widgets/communication/ui/communication.tsx:48]: If its needed, I can change it in the article, but why do we need communication-description, Where is it in the code?

File: src/widgets/communication/ui/communication.tsx (hunk 6 above)
Comment (ansivgit) [reply src/widgets/communication/ui/communication.tsx:44]: https://github.com/rolling-scopes/site/pull/536#discussion_r1773534157

File: src/widgets/communication/ui/communication.tsx (hunk 7 above)
Comment (ansivgit) [reply src/widgets/communication/ui/communication.tsx:48]: In code it is below, on UI here it is 
image

File: src/widgets/communication/ui/communication.module.scss (hunk 1 above)
[the remark this answers is not in the document]
Comment (thirdmadman) [reply src/widgets/communication/ui/communication.module.scss:32]: I dubble checked this - its brand color of discord
https://discord.com/branding

Considering this information, what should we do?

File: src/widgets/communication/ui/communication.tsx (hunk 6 above)
Comment (thirdmadman) [reply src/widgets/communication/ui/communication.tsx]: fixed: https://github.com/rolling-scopes/site/pull/536/commits/64ed4d8fac2ff1a8d9409660f2da96dc4bbe6be2

File: src/widgets/communication/ui/communication.tsx [hunk 8]
```diff
@@ -9,81 +10,63 @@ import {
...
-        <div className="column-2">
-          <figure className="disclogo-wrapper">
-            <DiscordLogo />
+    <section className={cx('container')}>
+      <article className={cx('content', 'communication-content')}>
+        <WidgetTitle mods="asterisk">{title}</WidgetTitle>
+        <div className={cx('communication-wrapper')}>
+          <figure className={cx('discord-logo-wrapper')}>
+            <Image src={discordLogo} alt="discord logo" />
           </figure>
```
[the remark this answers is not in the document]
Comment (thirdmadman) [reply src/widgets/communication/ui/communication.tsx:47]: Incorrect logic. Will ignore.

File: src/widgets/communication/ui/communication.test.tsx [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    it(`should render correct link of ${courseName}`, () => {
+      renderWithRouter(<Communication courseName={courseName as keyof typeof DISCORD_LINKS} />);
+      const linkElement = screen.getByText(/course discord server/i);
+
+      expect(linkElement).toBeVisible();
+      expect(linkElement.getAttribute('href')).toMatch(link);
+      cleanup();
+    });
+  });
+});
```
[the remark this answers is not in the document]
Comment (thirdmadman) [reply src/widgets/communication/ui/communication.test.tsx:63]: Ok, show me refactored code.

File: src/widgets/communication/ui/communication.tsx [hunk 10]
```diff
@@ -9,81 +10,63 @@
...
+    firstParagraphFirstHalf,
+    discordLink,
+    firstParagraphSecondHalf,
+    secondParagraphFirstHalf,
+    telegramLink,
+    secondParagraphSecondHalf,
+    thirdParagraphFirstHalf,
+    rsDocsLink,
+    thirdParagraphSecondHalf,
+  } = communicationText[lang];
```
[the remark this answers is not in the document]
Comment (thirdmadman) [reply src/widgets/communication/ui/communication.tsx:38]: Out of scope. Will ignore.

File: src/widgets/communication/ui/communication.test.tsx (hunk 3 above)
Comment (thirdmadman) [reply src/widgets/communication/ui/communication.test.tsx]: fiex: https://github.com/rolling-scopes/site/pull/536/commits/0d7d349989efc00fe3c38d8008c139313c543afa

File: src/widgets/communication/ui/communication.module.scss (hunk 1 above)
[the remark this answers is not in the document]
Comment (ansivgit) [reply src/widgets/communication/ui/communication.module.scss:32]: You can leave it in code in `hsl` format

[the remark this answers is not in the document]
Comment (thirdmadman) [reply src/widgets/communication/ui/communication.module.scss]: fixed: https://github.com/rolling-scopes/site/pull/536/commits/0d70007c1d92c7b8ad84753795c1157b79868fc8

**Review by ansivgit — APPROVED** [review]

**Pull request merged**